docs/source/developers/cpp/compute.rst

```diff
@@ -0,0 +1,171 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. highlight:: console
+.. _development-cpp-compute:
+
+============================
+Developing Arrow C++ Compute
+============================
+
+This section provides information for developers of Arrow C++ Compute module.
+
+Row Table
+=========
+
+The row table in Arrow represents data stored in row-major format. This format
+is particularly useful for scenarios involving random access to individual rows
+and where all columns are frequently accessed together. It is especially
+advantageous for hash-table keys and facilitates efficient operations such as
+grouping and hash joins by optimizing memory access patterns and data locality.
+
+Fixed-length vs. Varying-length
+-------------------------------
+
+A key property of the row table is whether it is fixed-length or varying-length.
+A fixed-length row table contains only fixed-length columns, while a
+varying-length row table includes at least one varying-length column. This
+distinction determines how data is stored and accessed in the row table.
+
+Buffer Layout
+-------------
+
+Similar to most Arrow ``Array``\s, the row table consists of three buffers:
+
+- **Null Masks Buffer**: Indicates null values for each column in each row.
+- **Fixed-length Buffer**: Stores row data for fixed-length tables or offsets to
+  varying-length data for varying-length tables.
+- **Varying-length Buffer** (Optional): Contains row data for varying-length
+  tables; unused for fixed-length tables.
+
+Row Format
+----------
+
+Null Masks
+~~~~~~~~~~
+
+For each row, a contiguous sequence of bits represents whether each column in
+that row is null. Each bit corresponds to a specific column. Unlike the
+convention in Arrow ``Array``\s, in a row table, a value of ``1`` indicates that
+the column is null, and a value of ``0`` indicates valid data. The null mask for
+a row occupies ``RowTableMetadata::null_masks_bytes_per_row`` bytes.
```

Comment:
```suggestion
For each row, a contiguous sequence of bits represents whether each column in
that row is null. Each bit corresponds to a specific column, with ``1``
indicating the value is null and ``0`` indicating the value is valid . Note that
this is the opposite of how the validity bitmap works for Arrays. The null mask
for a row occupies ``RowTableMetadata::null_masks_bytes_per_row`` bytes.
```

Style thing. I like to lead with the fact and qualify after. Feel free to ignore.

Reply:
Nice modification. Thanks!